### PR TITLE
Tests using scores

### PR DESCRIPTION
**Pull request by snath-xoc**
Preliminary check on using scoring (between predictions and y_test), instead of raw predicted values for conducting statistical test of equivalence. In a sense, this shifts from checking raw predictions to comparing the model's performance itself in distribution. Linked to discussion in #10 but compared first for transformers.

TO DO:

- [x] Try out other scores (so far have checked R2 and mean_squared_error)
- [x] Update handling of KMeans so that it is not pipelined into Ridge
- [x] AdaBoostClassifier amongst others are still returning 1 as p-values. I have implemented the shuffling for X_resampled which seems to give good enough results but maybe something more is needed, @jeremiedbb any ideas?
- [x] The test is still a bit finicky, sometimes 15 estimators fail the statistical tests and sometimes 12, this may just be random fluke.
- [ ] check for deterministic using the predictions (not scores)
- [ ] Add flag if deterministic is detected but continue to use kstest

### CODE REVIEW

**Comment by snath-xoc** [comment]
@ogrisel I tried checking equivalence on the score calculated across predictions (using the ridge regression score attribute that calculates the R2) for MiniBatchKMeans (that we know is currently wrong). The p-value however is:

```
EquivalenceTestResult(estimator_name='MiniBatchKMeans', test_name='kstest', p_value=0.2059692354005536)
```

As compared to what we get when comparing the raw predicted values in distribution

```
EquivalenceTestResult(estimator_name='MiniBatchKMeans', test_name='kstest', min_p_value=0.005536865013092584, mean_p_value=0.6344460028520363)
```

Apart from this the NoisyTransformer test passes, so it seems like the score is a bit underpowered in some cases.

**Comment by ogrisel** [comment]
> Apart from this the NoisyTransformer test passes, so it seems like the score is a bit underpowered in some cases?

Or maybe it's just bad luck. If there is no bug in the sample weight handling of `RandomTreesEmbedding`, it's expected to get uniformly distributed pvalues between 0 and 1 (for different random state of the outer testing procedure).

Maybe we can test on other transformers, with known bugs such as `KBinsDiscretizer` in `main` vs your PR.

File: src/sample_weight_audit/estimator_check.py [hunk 1]
```diff
@@ -163,7 +170,7 @@ def get_cv_params(
     return extra_params_weighted, extra_params_repeated
 
 
-def compute_predictions(est, X):
+def compute_predictions(est, X, y=None, score=True):
```
Comment (ogrisel) [inline src/sample_weight_audit/estimator_check.py:173]: This is no longer computing predictions. We need to always adjust function names and variable names to reflect what the code is doing, otherwise the code will quickly become unmaintainable.

If we want to keep the `score=False` option, let's instead write 2 functions: `compute_predictions(est, X)` and another called `score_estimator(est, X, y)` and the latter can call the former internally to avoid code duplication if needed.

File: src/sample_weight_audit/estimator_check.py [hunk 2]
```diff
@@ -172,12 +179,37 @@ def compute_predictions(est, X):
             return est.predict_proba(X)
         else:
             return est.decision_function(X)
+    elif isinstance(est, Pipeline):
+        if score:
+            return mean_squared_error(est.predict(X), y)
```
Comment (ogrisel) [inline src/sample_weight_audit/estimator_check.py:184]: This is only valid for regressors and should therefore go into the regressor branch, irrespective of whether the estimator is a pipeline or not.

For classifiers, we should leverage maximally informative, threshold independent classification metrics such as `log_loss` on the output of `predict_proba` and `roc_auc_score` on the output of `decision_function`.

Comment (ogrisel) [reply src/sample_weight_audit/estimator_check.py:184]: And for clusterers, we should use clustering metrics that are invariant to label permutations (e.g. Rand index, NMI, V-measure and the like).

File: src/sample_weight_audit/estimator_check.py [hunk 3]
```diff
@@ -108,7 +110,6 @@ def check_weighted_repeated_estimator_fit_equivalence(
     if np.all(diffs < np.finfo(diffs.dtype).eps):
         raise UnexpectedDeterministicPredictions(message)
 
-
     assert predictions_weighted.ndim == 3
```
Comment (ogrisel) [inline src/sample_weight_audit/estimator_check.py:113]: If we switch to scores, we should do it for all estimators (classifiers, regressors, transformers, clusterers...) and never run the tests on nd predictions but always on 1d score values:

We would always run `run_1d_test(scores_weighted, scores_repeated)`  `scores_weighted` and `scores_repeated` would be 1d arrays with shape `(n_stochastic_fits,)` for all estimators, and we can remove the aggregation logic of pvalues.

We should also simplify the code that handles the relationship between `stat_test_dim` and `prediction_dim`: we can actually remove the `stat_test_dim` argument altogether.

Furthermore, we can also get rid of the `get_diverse_subset` function: we should instead compute scores on large enough test sets (e.g. 1000 data points or more, that directly result from the train/test split without diverse subset extraction).

All in all, this would simplify the code a lot and make it easier to homogeneously support clusterers and transformers alongside classifiers and regressors.

**Comment by ogrisel** [comment]
I merged the other PR. There is now a conflict to solve but it's happening on a line where I put a review comment above (https://github.com/snath-xoc/sample-weight-audit-nondet/pull/13#discussion_r1946714472) so please take that into account when syncing this branch with `main`.

**Comment by snath-xoc** [comment]
> > Apart from this the NoisyTransformer test passes, so it seems like the score is a bit underpowered in some cases?
> 
> It's possible. Or maybe it's just bad luck. If there is no bug in the sample weight handling of `RandomTreesEmbedding`, it's expected to get uniformly distributed pvalues between 0 and 1 (for different random state of the outer testing procedure).
> 
> Maybe we can test on other transformers, with known bugs such as `KBinsDiscretizer` in `main` vs your PR.

This still needs to be done, the main branch of KBins had the bug where sample weights were not resampled correctly

**Comment by ogrisel** [comment]
> however MinibatchKMeans is also unexpectedly passing

Since we merged the transformer pipeline trick merged in #12, it's indeed possible that our strategy is now less sensitive because the 2 ridge models (repeated vs weighted) can compensate for some of the discrepancies between the extracted (repeated vs weighted) transformed generated feature sets.

I have an idea to avoid this problem. I will open a dedicated PR as follow-up to #12 because it should be mergeable independently of #13.

File: src/sample_weight_audit/estimator_check.py [hunk 4]
```diff
@@ -80,65 +73,52 @@ def check_weighted_repeated_estimator_fit_equivalence(
...
+    assert scores_repeated.shape == scores_weighted.shape
 
-    data_to_test_weighted = predictions_weighted.reshape(
-        (n_stochastic_fits, stat_test_dim)
-    ).T  # shape: (stat_test_dim, n_stochastic_fits)
-    data_to_test_repeated = predictions_repeated.reshape(
-        (n_stochastic_fits, stat_test_dim)
-    ).T  # shape: (stat_test_dim, n_stochastic_fits)
+    data_to_test_weighted = scores_weighted.T  # shape: (1, n_stochastic_fits)
+    data_to_test_repeated = scores_repeated.T  # shape: (1, n_stochastic_fits)
```
Comment (ogrisel) [inline src/sample_weight_audit/estimator_check.py:108]: We should further simplify this code to make it such that `scores_repeated` and `scores_weighted` are directly constructed by `multifit_over_weighted_and_repeated` as 1d arrays with shape `(n_stochastic_fits,)`.

File: src/sample_weight_audit/estimator_check.py [hunk 5]
```diff
@@ -178,6 +158,26 @@ def compute_predictions(est, X):
...
+        print("y-values are required when score is set to True")
+    if is_regressor(est):
+        return mean_squared_error(preds, y)
+    elif is_classifier(est):
+        if hasattr(est, "predict_proba"):
+            return log_loss(y, preds)
+        else:
+            return roc_auc_score(preds, y)
+    elif hasattr(est, "transform"):
+        return rand_score(preds, y)
```
Comment (ogrisel) [inline src/sample_weight_audit/estimator_check.py:176]: We should use `rand_score` in estimators that are clusterers (detected with `is_clusterer`), not on all transformers.

```suggestion
    elif is_clusterer(est):
        return rand_score(preds, y)
```

Transformers (that are not clusterers) should instead be pipelined with a ridge regressor and therefore treated as regressors.

Comment (ogrisel) [reply src/sample_weight_audit/estimator_check.py:176]: Similarly, `check_pipeline_and_fit` should be updated to avoid pipelining clusterers.

**Review by ogrisel — COMMENTED** [review]
Let's try to see if we can make the test faire for MiniBatchKMeans when treating it as a clusterer (and therefore using the Rand index) instead of a transformer. See below for details:

File: src/sample_weight_audit/estimator_check.py [hunk 6]
```diff
@@ -178,6 +158,26 @@ def compute_predictions(est, X):
         raise NotImplementedError(f"Estimator type not supported: {est}")
 
 
+def score_estimator(est, X, y=None, score=True):
+
+    preds = compute_predictions(est, X)
+    if not score:
+        return preds
```
[the remark this answers is not in the document]
Comment (ogrisel) [reply src/sample_weight_audit/estimator_check.py:165]: See: https://en.wikipedia.org/wiki/You_aren%27t_gonna_need_it (YAGNI).

**Comment by snath-xoc** [comment]
Oh no, yes I had that issue initially but found setting fit_intercept to false helped at least in the tests.
> EDIT: the idea I had in mind (share a common ridge instance in the two pipelines) happens to be wrong.

File: src/sample_weight_audit/data.py [hunk 7]
```diff
@@ -30,9 +18,15 @@ def weighted_and_repeated_train_test_split(
...
     repeated_indices = np.repeat(np.arange(X_train.shape[0]), sample_weight_train)
-    X_resampled_by_weights = np.take(X_train, repeated_indices, axis=0)
-    y_resampled_by_weights = np.take(y_train, repeated_indices, axis=0)
+    shuffled_indices = rng.permutation(len(repeated_indices))
+    X_resampled_by_weights = np.take(X_train, repeated_indices, axis=0)[
+        shuffled_indices
+    ]
+    y_resampled_by_weights = np.take(y_train, repeated_indices, axis=0)[
+        shuffled_indices
+    ]
```
Comment (snath-xoc) [inline src/sample_weight_audit/data.py:30]: This implements what was suggested in Issue #14 @ogrisel and @jeremiedbb please check. It does work with KMeans on my end though (getting p-values of 0.21)

File: src/sample_weight_audit/estimator_check.py (hunk 4 above)
Comment (snath-xoc) [reply src/sample_weight_audit/estimator_check.py]: Test now asserts that scores_weighted/repeated.ndim==1 and we keep dimesnion as 1-D after stacking.

File: src/sample_weight_audit/data.py [hunk 8]
```diff
@@ -30,9 +18,15 @@ def weighted_and_repeated_train_test_split(
             X, y, sample_weight, train_size=train_size, random_state=random_state
         )
     )
+    rng = np.random.default_rng(random_state)
     repeated_indices = np.repeat(np.arange(X_train.shape[0]), sample_weight_train)
-    X_resampled_by_weights = np.take(X_train, repeated_indices, axis=0)
-    y_resampled_by_weights = np.take(y_train, repeated_indices, axis=0)
+    shuffled_indices = rng.permutation(len(repeated_indices))
```
[the remark this answers is not in the document]
Comment (snath-xoc) [reply src/sample_weight_audit/data.py:23]: Have added this now under a separate PR #16

File: src/sample_weight_audit/estimator_check.py [hunk 9]
```diff
@@ -163,23 +141,32 @@ def get_cv_params(
...
+            return log_loss(y, preds)
         else:
-            return est.decision_function(X)
-    elif hasattr(est, "transform"):
-        return est.transform(X)
+            preds = est.decision_function(X)
+            return roc_auc_score(preds, y)
+    elif is_clusterer(est):
+        preds = est.predict(X)
+        return rand_score(preds, y)
```
[the remark this answers is not in the document]
Comment (snath-xoc) [reply src/sample_weight_audit/estimator_check.py]: Hmmmm so I incorporated this, but now KMeans is giving failing the deterministic check (i.e., scores from weighted and repeated are exactly the same within numerical precision)

**Comment by snath-xoc** [comment]
Merged by mistake undoing that now.
